### PR TITLE
LOOP-2339: Nov DIY merge

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-2339

* Travis build changes: update to xcode12.2, and cache carthage builds
* Starting values when entering new therapy settings from scratch 
* Mapping model to migrate from v1 (current DIY release) to v3 model.
* Parameterize app name
* Updates to ServiceUI protocol to support onboarding, and additional support link.
* iOS 14 datepicker updates to use previous style
* Use preferred units on initial settings creation
* Use framework strings for localization
* Fix iOS 14 issue with schedule editors not being dismissed: https://tidepool.atlassian.net/browse/LOOP-2122

### CODE REVIEW

File: Extensions/Guardrail+Settings.swift [hunk 1]
```diff
@@ -23,7 +23,7 @@ public extension Guardrail where Value == HKQuantity {
         .min()!
     }
 
-    static let correctionRange = Guardrail(absoluteBounds: 87...180, recommendedBounds: 101...115, unit: .milligramsPerDeciliter)
+    static let correctionRange = Guardrail(absoluteBounds: 87...180, recommendedBounds: 101...115, unit: .milligramsPerDeciliter, startingSuggestion: 100)
```
Comment (nhamming) [inline Extensions/Guardrail+Settings.swift]: Why is the starting suggestion outside the recommended bounds?

File: LoopKit/TherapySettings.swift [hunk 2]
```diff
@@ -34,6 +34,20 @@ public struct TherapySettings: Equatable, Codable {
         return glucoseTargetRangeSchedule?.unit
     }
     
+    public var isComplete: Bool {
+        return
+            glucoseTargetRangeSchedule != nil &&
+            //preMealTargetRange != nil &&
+            //workoutTargetRange != nil &&
```
Comment (nhamming) [inline LoopKit/TherapySettings.swift:42]: [nit] remove?

File: LoopKitUI/Views/Settings Editors/TherapySettingsView.swift [hunk 3]
```diff
@@ -183,11 +184,11 @@ extension TherapySettingsView {
     
     private var preMealCorrectionRangeSection: some View {
         section(for: .preMealCorrectionRangeOverride) {
-            if self.glucoseUnit != nil && self.viewModel.therapySettings.glucoseTargetRangeSchedule != nil {
+            if let glucoseUnit = self.glucoseUnit, let schedule = self.viewModel.therapySettings.glucoseTargetRangeSchedule {
```
Comment (nhamming) [inline LoopKitUI/Views/Settings Editors/TherapySettingsView.swift:187]: use `schedule` on line 194?

File: LoopKitUI/Views/Settings Editors/TherapySettingsView.swift [hunk 4]
```diff
@@ -198,11 +199,11 @@ extension TherapySettingsView {
     
     private var workoutCorrectionRangeSection: some View {
         section(for: .workoutCorrectionRangeOverride) {
-            if self.glucoseUnit != nil && self.viewModel.therapySettings.glucoseTargetRangeSchedule != nil {
+            if let glucoseUnit = self.glucoseUnit, self.viewModel.therapySettings.glucoseTargetRangeSchedule != nil {
```
Comment (nhamming) [inline LoopKitUI/Views/Settings Editors/TherapySettingsView.swift]: user `schedule` on line 209?

**Review by nhamming — APPROVED** [review]
Just non-blocking questions/suggestions. LGTM

File: Extensions/Guardrail+Settings.swift (hunk 1 above)
Comment (ps2) [reply Extensions/Guardrail+Settings.swift]: Two things: 1) The Tidepool bounds are being updated to start at 100, instead of 101. 2) The starting suggestion is really only used in DIY, since Tidepool users will be starting with their provider's recommendation. 

Also, Guardrails are something that are likely to diverge at some point, so DIY will be able to have different settings than us.

File: LoopKit/TherapySettings.swift (hunk 2 above)
Comment (ps2) [reply LoopKit/TherapySettings.swift:42]: I kept those there as comments that they're purposefully not included in the check.

File: Extensions/Guardrail+Settings.swift (hunk 1 above)
Comment (darinkrauss) [reply Extensions/Guardrail+Settings.swift]: Should we just update the recommended bounds to 100 here, then?

File: Extensions/Guardrail+Settings.swift [hunk 5]
```diff
@@ -92,7 +94,8 @@ public extension Guardrail where Value == HKQuantity {
         return Guardrail(
             absoluteBounds: allowedBasalRates.first!...allowedBasalRates.last!,
             recommendedBounds: allowedBasalRates.first!...allowedBasalRates.last!,
-            unit: .internationalUnitsPerHour
+            unit: .internationalUnitsPerHour,
+            startingSuggestion: 0
```
Comment (darinkrauss) [inline Extensions/Guardrail+Settings.swift]: Should the `startingSuggestion` be `allowedBasalRates.first!` instead of `0`?

File: Extensions/Guardrail+Settings.swift [hunk 6]
```diff
@@ -127,7 +130,8 @@ public extension Guardrail where Value == HKQuantity {
             return Guardrail(
                 absoluteBounds: supportedBasalRates.first!...absoluteUpperBound,
                 recommendedBounds:  supportedBasalRates.first!...absoluteUpperBound,
-                unit: .internationalUnitsPerHour
+                unit: .internationalUnitsPerHour,
+                startingSuggestion: 3
```
Comment (darinkrauss) [inline Extensions/Guardrail+Settings.swift]: Should the `startingSuggestion` be `3.clamped(to: supportedBasalRates.first!...absoluteUpperBound)`?

File: Extensions/Guardrail+Settings.swift [hunk 7]
```diff
@@ -140,7 +144,8 @@ public extension Guardrail where Value == HKQuantity {
         return Guardrail(
             absoluteBounds: supportedBolusVolumes.first!...supportedBolusVolumes.last!,
             recommendedBounds: supportedBolusVolumes.dropFirst().first!...recommendedUpperBound!,
-            unit: .internationalUnit()
+            unit: .internationalUnit(),
+            startingSuggestion: 5
```
Comment (darinkrauss) [inline Extensions/Guardrail+Settings.swift]: Should the `startingSuggestion` be `5.clamped(to: supportedBolusVolumes.dropFirst().first!...recommendedUpperBound!)`?

File: Extensions/Guardrail+Settings.swift (hunk 6 above)
Comment (darinkrauss) [reply Extensions/Guardrail+Settings.swift]: If so, also consider pulling out `supportedBasalRates.first!...absoluteUpperBound` into a local variable.

File: Extensions/Guardrail+Settings.swift (hunk 7 above)
Comment (darinkrauss) [reply Extensions/Guardrail+Settings.swift]: If so, also consider pulling out `supportedBolusVolumes.dropFirst().first!...recommendedUpperBound!` into a local variable.

File: LoopKit/CarbKit/CarbStore.swift [hunk 8]
```diff
@@ -220,7 +220,7 @@ public final class CarbStore: HealthKitSampleStore {
         self.provenanceIdentifier = provenanceIdentifier
         
         let observationEnabled = observationInterval > 0
-
+
```
Comment (darinkrauss) [inline LoopKit/CarbKit/CarbStore.swift]: [nit] Consider backing out only whitespace change since it is the only change in this file.

File: LoopKit/Guardrail.swift [hunk 9]
```diff
@@ -24,12 +24,14 @@ public enum SafetyClassification: Equatable {
 public struct Guardrail<Value: Comparable> {
     public let absoluteBounds: ClosedRange<Value>
     public let recommendedBounds: ClosedRange<Value>
+    public let startingSuggestion: Value?
 
-    public init(absoluteBounds: ClosedRange<Value>, recommendedBounds: ClosedRange<Value>) {
+    public init(absoluteBounds: ClosedRange<Value>, recommendedBounds: ClosedRange<Value>, startingSuggestion: Value? = nil) {
         precondition(absoluteBounds.lowerBound <= recommendedBounds.lowerBound, "The minimum value must be less than or equal to the smallest recommended value")
         precondition(absoluteBounds.upperBound >= recommendedBounds.upperBound, "The maximum value must be greater than or equal to the greatest recommended value")
```
Comment (darinkrauss) [inline LoopKit/Guardrail.swift:31]: Should there be a precondition that `startingSuggestion` is within `absoluteBounds`?

File: LoopKit/LoopPluginBundleKey.swift [hunk 10]
```diff
@@ -16,4 +16,5 @@ public enum LoopPluginBundleKey: String {
     case serviceDisplayName = "com.loopkit.Loop.ServiceDisplayName"
     case serviceIdentifier = "com.loopkit.Loop.ServiceIdentifier"
     case extensionIdentifier = "com.loopkit.Loop.ExtensionIdentifier"
+    case providesOnboarding = "com.loopkit.Loop.Plugin.ProvidesOnboarding"
```
Comment (darinkrauss) [inline LoopKit/LoopPluginBundleKey.swift:19]: [nit] Why the addition of the `.Plugin.`component?

File: LoopKit/TherapySettings.swift (hunk 2 above)
Comment (darinkrauss) [reply LoopKit/TherapySettings.swift:42]: [nit] Consider changing to or adding a descriptive so this is clear. (To me it looks like these were temporarily removed for testing or such.)

File: LoopKitUI/Views/GuardrailWarning.swift [hunk 11]
```diff
@@ -79,7 +79,7 @@ public struct GuardrailWarning: View {
                 }
             } else {
                 // ANNA TODO: figure out this one
-                return Text("Some of the values you have entered are outside of what Tidepool typically recommends for most people.")
+                return Text(LocalizedString("Some of the values you have entered are outside of what is typically recommended for most people.", comment: "Caption for guardrail warning when more than one threshold is crossed"))
```
Comment (darinkrauss) [inline LoopKitUI/Views/GuardrailWarning.swift:82]: [nit] Will removal of "Tidepool" be okay for Tidepool Loop?

File: LoopKitUI/Views/Settings Editors/CarbRatioScheduleEditor.swift [hunk 12]
```diff
@@ -61,7 +62,7 @@ public struct CarbRatioScheduleEditor: View {
             guardrail: .carbRatio,
             selectableValueStride: HKQuantity(unit: .realCarbRatioScheduleUnit, doubleValue: 0.01),
             quantitySelectionMode: .fractional,
-            defaultFirstScheduleItemValue: Guardrail.carbRatio.absoluteBounds.upperBound,
+            defaultFirstScheduleItemValue: Guardrail.carbRatio.startingSuggestion ?? Guardrail.carbRatio.absoluteBounds.upperBound,
```
Comment (darinkrauss) [inline LoopKitUI/Views/Settings Editors/CarbRatioScheduleEditor.swift]: [nit] I notice the other code with this pattern uses the `recommendedBounds.upperBound` and not `absoluteBounds.upperBound`. Consider changing this for uniformity.

File: LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift [hunk 13]
```diff
@@ -211,7 +209,8 @@ public struct InsulinModelSelection: View, HorizontalSizeClassOverride {
     var insulinModelSettingDescription: Text {
         let spellOutFormatter = NumberFormatter()
         spellOutFormatter.numberStyle = .spellOut
-        return Text("\(appName) assumes insulin is actively working for 6 hours. You can choose from \(selectableInsulinModelSettings.count as NSNumber, formatter: spellOutFormatter) different models for how the app measures the insulin’s peak activity.", comment: "Insulin model setting description (1: app name) (2: number of models)")
+        let modelCountString = spellOutFormatter.string(from: selectableInsulinModelSettings.count as NSNumber)!
+        return Text(String(format: LocalizedString("%1$@ assumes insulin is actively working for 6 hours. You can choose from %2$@ different models for how the app measures the insulin’s peak activity.", comment: "Insulin model setting description (1: app name) (2: number of models)"), appName, modelCountString))
```
Comment (darinkrauss) [inline LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:213]: [nit] Technically if there is only 1 to choose from it reads "can choose from 1 different models", but there is no real choice (and plurality is off).

File: LoopKitUI/Views/Settings Editors/InsulinSensitivityScheduleEditor.swift [hunk 14]
```diff
@@ -59,7 +60,7 @@ public struct InsulinSensitivityScheduleEditor: View {
             unit: sensitivityUnit,
             guardrail: .insulinSensitivity,
             selectableValueStride: stride,
-            defaultFirstScheduleItemValue: Guardrail.insulinSensitivity.absoluteBounds.upperBound,
+            defaultFirstScheduleItemValue: Guardrail.insulinSensitivity.startingSuggestion ?? Guardrail.insulinSensitivity.absoluteBounds.upperBound,
```
Comment (darinkrauss) [inline LoopKitUI/Views/Settings Editors/InsulinSensitivityScheduleEditor.swift]: [nit] Other code uses `recommendedBounds.upperBound`. Consider switching here, too.

File: LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift [hunk 15]
```diff
@@ -57,9 +57,11 @@ public struct SuspendThresholdEditor: View {
                 workoutTargetRange: viewModel.therapySettings.workoutTargetRange?.quantityRange(for: unit)
             ),
             onSave: { [weak viewModel] newValue in
-                let newThreshold = GlucoseThreshold(unit: viewModel!.therapySettings.glucoseUnit!, value: newValue.doubleValue(for: viewModel!.therapySettings.glucoseUnit!))
-                viewModel?.saveSuspendThreshold(value: newThreshold)
-                didSave?()
+                if let viewModel = viewModel {
```
Comment (darinkrauss) [inline LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift]: [nit] Consider `guard let viewModel = viewModel else { return }` for more similarity with other "weak" checks in the code.

File: LoopKitUI/Views/Settings Editors/TherapySettingsView.swift [hunk 16]
```diff
@@ -300,12 +301,14 @@ extension TherapySettingsView {
...
     private var insulinSensitivitiesSection: some View {
         section(for: .insulinSensitivity) {
-            if self.viewModel.therapySettings.insulinSensitivitySchedule != nil && self.sensitivityUnit != nil {
-                ForEach(self.viewModel.therapySettings.insulinSensitivitySchedule!.items, id: \.self) { value in
-                    ScheduleValueItem(time: value.startTime,
-                                      value: value.value,
-                                      unit: self.sensitivityUnit!,
-                                      guardrail: Guardrail.insulinSensitivity)
+            if let sensitivityUnit = self.sensitivityUnit, let schedule = self.viewModel.therapySettings.insulinSensitivitySchedule {
+                if self.viewModel.therapySettings.insulinSensitivitySchedule != nil {
```
Comment (darinkrauss) [inline LoopKitUI/Views/Settings Editors/TherapySettingsView.swift]: [nit] It looks like this `if self.viewModel.therapySettings.insulinSensitivitySchedule != nil` is unnecessary due to line above?

File: LoopKitUI/ServiceUI.swift [hunk 17]
```diff
@@ -8,28 +8,39 @@
...
     static var image: UIImage? { get }
+    
+    /// Indicates whether this service provides onboarding (configuring therapy settings)
+    static var providesOnboarding: Bool { get }
 
     /// Provides a view controller to create and configure a new service, if needed.
     ///
     /// - Returns: A view controller to create and configure a new service.
-    static func setupViewController() -> (UIViewController & ServiceSetupNotifying & CompletionNotifying)?
+    static func setupViewController(currentTherapySettings: TherapySettings, preferredGlucoseUnit: HKUnit, chartColors: ChartColorPalette, carbTintColor: Color, glucoseTintColor: Color, guidanceColors: GuidanceColors, insulinTintColor: Color) -> (UIViewController & ServiceSetupNotifying & CompletionNotifying)?
```
Comment (darinkrauss) [inline LoopKitUI/ServiceUI.swift:31]: Not sure I understand why therapy settings and preferred glucose unit would be passed to service setup/settings UI. Later we might want to revisit the "service" concept and how that interacts with therapy settings (if it does). In particular, I'm thinking of the TidepoolService being used in DIY.

File: LoopKitUI/ServiceUI.swift [hunk 18]
```diff
@@ -8,28 +8,39 @@
 
 import LoopKit
 import SwiftUI
+import HealthKit
+
+public protocol SupportInfoProvider {
```
Comment (darinkrauss) [inline LoopKitUI/ServiceUI.swift:13]: Also, not sure this belongs in "service". Perhaps we should consider breaking out these "service" concepts and keeping them all separate. Later, though.

File: LoopKit/InsulinKit/DoseStore.swift [hunk 19]
```diff
@@ -1125,7 +1125,7 @@ extension DoseStore {
             case .failure(let error):
                 completion(.failure(.persistenceError(description: error.localizedDescription, recoverySuggestion: nil)))
             case .success(let insulinDeliveryDoses):
-                let filteredStart = max(insulinDeliveryDoses.lastBasalEndDate ?? start, start)
+                let filteredStart = max(self.lastPumpEventsReconciliation ?? start, start)
```
Comment (darinkrauss) [inline LoopKit/InsulinKit/DoseStore.swift:1128]: Why this change? I checked the code before my large DoseStore update and it used `insulinDeliveryDoses.lastBasalEndDate`, too.

**Review by darinkrauss — APPROVED** [review]
A number of comments and a couple questions, but nothing show stopping. LGTM!

File: LoopKit/TherapySettings.swift (hunk 2 above)
Comment (rickpasetto) [reply LoopKit/TherapySettings.swift:42]: +1 to Darin's comment.

File: LoopKitUI/ServiceUI.swift (hunk 18 above)
Comment (rickpasetto) [reply LoopKitUI/ServiceUI.swift:13]: +1

File: LoopKitUI/Views/ConfigurationPage.swift [hunk 20]
```diff
@@ -87,7 +87,7 @@ extension ConfigurationPage {
     ) {
         self.init(
             title: title,
-            actionButtonTitle: Text("Save", comment: "The button text for saving on a configuration page"),
+            actionButtonTitle: Text(LocalizedString("Save", comment: "The button text for saving on a configuration page")),
```
Comment (rickpasetto) [inline LoopKitUI/Views/ConfigurationPage.swift:90]: Why is this correct and the previous form not correct?

File: LoopKitUI/Views/GuardrailWarning.swift (hunk 11 above)
Comment (rickpasetto) [reply LoopKitUI/Views/GuardrailWarning.swift:82]: This will need to be approved by Melissa Lee.

File: LoopKitUI/Views/LabeledNumberInput.swift [hunk 21]
```diff
@@ -52,7 +52,7 @@ public struct LabeledNumberInput: View {
                     .multilineTextAlignment(.trailing)
                     .keyboardType(self.allowFractions ? .decimalPad : .numberPad)
                     .frame(width: geometry.size.width/2, alignment: .trailing)
-                    .accessibility(label: Text("Enter value"))
+                    .accessibility(label: Text(String(format: LocalizedString("Enter %1$@ value", comment: "Format string for accessibility label for value entry. (1: value label)"), label)))
```
Comment (rickpasetto) [inline LoopKitUI/Views/LabeledNumberInput.swift:55]: [nit] will this affect the detox UI tests?

File: Extensions/Guardrail+Settings.swift [hunk 22]
```diff
@@ -9,7 +9,7 @@
 import HealthKit
 
 public extension Guardrail where Value == HKQuantity {
-    static let suspendThreshold = Guardrail(absoluteBounds: 67...110, recommendedBounds: 74...80, unit: .milligramsPerDeciliter)
+    static let suspendThreshold = Guardrail(absoluteBounds: 67...110, recommendedBounds: 74...80, unit: .milligramsPerDeciliter, startingSuggestion: 80)
```
Comment (rickpasetto) [inline Extensions/Guardrail+Settings.swift:12]: Please have these `startingSuggestion`s reviewed by Product.

File: LoopKitUI/Views/Settings Editors/CorrectionRangeScheduleEditor.swift [hunk 23]
```diff
@@ -40,14 +42,13 @@ public struct CorrectionRangeScheduleEditor: View {
...
-        precondition(viewModel.therapySettings.glucoseUnit != nil)
         self.init(
             schedule: viewModel.therapySettings.glucoseTargetRangeSchedule,
-            unit: viewModel.therapySettings.glucoseUnit!,
+            unit: viewModel.therapySettings.glucoseTargetRangeSchedule?.unit ?? viewModel.preferredGlucoseUnit,
             minValue: Guardrail.minCorrectionRangeValue(suspendThreshold: viewModel.therapySettings.suspendThreshold),
             onSave: { [weak viewModel] newSchedule in
-                viewModel?.saveCorrectionRange(range: newSchedule)
                 didSave?()
+                viewModel?.saveCorrectionRange(range: newSchedule)
```
Comment (rickpasetto) [inline LoopKitUI/Views/Settings Editors/CorrectionRangeScheduleEditor.swift:49]: I don't think this is correct. `didSave` is intentionally called _after_ the viewModel saves the data.

File: LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift [hunk 24]
```diff
@@ -62,8 +62,17 @@ public struct DeliveryLimitsEditor: View {
            didSave: (() -> Void)? = nil
     ) {
         precondition(viewModel.pumpSupportedIncrements != nil)
-        let maxBasal = HKQuantity(unit: .internationalUnitsPerHour, doubleValue: viewModel.therapySettings.maximumBasalRatePerHour!)
-        let maxBolus = HKQuantity(unit: .internationalUnit(), doubleValue: viewModel.therapySettings.maximumBolus!)
+        
+        var maxBasal: HKQuantity?
+        if let maximumBasalRatePerHour = viewModel.therapySettings.maximumBasalRatePerHour {
```
Comment (rickpasetto) [inline LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift]: [nit] consider instead:
```swift
        let maxBasal = viewModel.therapySettings.maximumBasalRatePerHour.map { HKQuantity(unit: .internationalUnitsPerHour, doubleValue: $0) }
```

File: LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift [hunk 25]
```diff
@@ -62,8 +62,17 @@ public struct DeliveryLimitsEditor: View {
...
+            maxBasal = HKQuantity(unit: .internationalUnitsPerHour, doubleValue: maximumBasalRatePerHour)
+        }
+
+        var maxBolus: HKQuantity?
```
Comment (rickpasetto) [inline LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift]: [nit] consider instead:
```swift
        let maxBolus = viewModel.therapySettings.maximumBolus.map { HKQuantity(unit: . internationalUnit(), doubleValue: $0) }
```

File: LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift (hunk 15 above)
Comment (rickpasetto) [reply LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift]: +1

File: LoopKitUI/Views/Settings Editors/TherapySettingsView.swift [hunk 26]
```diff
@@ -152,12 +153,12 @@ extension TherapySettingsView {
     
     private var suspendThresholdSection: some View {
         section(for: .suspendThreshold, header: viewModel.prescription == nil ? AnyView(Spacer()) : AnyView(EmptyView())) {
-            if self.glucoseUnit != nil {
+            if let glucoseUnit = self.glucoseUnit {
```
Comment (rickpasetto) [inline LoopKitUI/Views/Settings Editors/TherapySettingsView.swift:156]: Hooray they finally added this feature to **Swift** 😄

File: LoopKitUI/Views/Settings Editors/TherapySettingsView.swift (hunk 16 above)
Comment (rickpasetto) [reply LoopKitUI/Views/Settings Editors/TherapySettingsView.swift]: +1 please remove.

**Review by rickpasetto — CHANGES_REQUESTED** [review]
I have a couple of small changes requested, plus a request to review Guardrail "suggested values" with Product.

File: Extensions/Guardrail+Settings.swift (hunk 1 above)
Comment (ps2) [reply Extensions/Guardrail+Settings.swift]: Rick was already tackling that, and I think it's already landed, and will be merged in before this lands.

File: LoopKit/Guardrail.swift (hunk 9 above)
Comment (ps2) [reply LoopKit/Guardrail.swift:31]: I think these are judgement calls, and generally don't add more code if don't absolutely need to (I tend to assign a high cost to a line of code, multiplied by all the future readers). But I'm happy to do so where people think it is worth it.

File: LoopKit/LoopPluginBundleKey.swift (hunk 10 above)
Comment (ps2) [reply LoopKit/LoopPluginBundleKey.swift:19]: Because `com.loopkit.Loop.ProvidesOnboarding` doesn't seem scoped enough.

File: LoopKitUI/Views/GuardrailWarning.swift (hunk 11 above)
Comment (ps2) [reply LoopKitUI/Views/GuardrailWarning.swift:82]: We have discussed this several times. Tidepool specific text in code common to Tidepool Loop and DIY needs to wait until we have the facility for Tidepool Loop string substitution, or it will block merges.

File: LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift (hunk 13 above)
Comment (ps2) [reply LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:213]: We don't need to support only one choice for either DIY or Tidepool Loop.

File: LoopKitUI/ServiceUI.swift (hunk 17 above)
Comment (ps2) [reply LoopKitUI/ServiceUI.swift:31]: Services may provide onboarding now, and thus need to know preferred glucose units and current therapy settings to do so.

File: LoopKitUI/ServiceUI.swift (hunk 18 above)
Comment (ps2) [reply LoopKitUI/ServiceUI.swift:13]: Yeah, I could see some refactoring of these responsibilities out Type-wise; they do have strong relationships with the services (settings being stored in tidepool/nightscout backend, support being handled by tidepool backend, etc.) that would still exist, but I'm sure there are improvements that could be made to the type system. I do think this should be done as a later refactoring, though, if we think that functionally this is not problematic.

File: LoopKit/InsulinKit/DoseStore.swift (hunk 19 above)
Comment (ps2) [reply LoopKit/InsulinKit/DoseStore.swift:1128]: This is a fix for a pretty severe issue where Loop displays double IOB at times, if a pump reports reservoir more recently than a finalized basal dose. Mostly an issue for DIY MDT; should not affect pods. See https://github.com/LoopKit/LoopKit/pull/344

File: LoopKitUI/Views/ConfigurationPage.swift (hunk 20 above)
Comment (ps2) [reply LoopKitUI/Views/ConfigurationPage.swift:90]: The previous form will look up translations in the app's bundle. The LocalizedString method is a helper that looks in the local framework's bundle for translations.

File: Extensions/Guardrail+Settings.swift (hunk 22 above)
Comment (ps2) [reply Extensions/Guardrail+Settings.swift:12]: These starting suggestions do not affect Tidepool Loop, as the onboarding will use initial values from the user's prescription. If Tidepool Loop eventually allows clean slate setup, we will need to tackle providing these. At some point guardrails and initial recommendations will likely diverge between Tidepool Loop and DIY, and we'll need to parameterize it in some fashion.

File: LoopKitUI/Views/Settings Editors/CorrectionRangeScheduleEditor.swift (hunk 23 above)
Comment (ps2) [reply LoopKitUI/Views/Settings Editors/CorrectionRangeScheduleEditor.swift:49]: Ah, this was leftover from experiments of trying to fix the non-dismissing schedules in iOS 14 (didSave was the dismiss call). I reverted it, and my other fix (dispatch async) still works. Will revert this.

File: LoopKit/Guardrail.swift (hunk 9 above)
Comment (darinkrauss) [reply LoopKit/Guardrail.swift:31]: Ultimately up to you. Given that there are already two preconditions to ensure the first two parameters make sense, I don't think a third precondition validating the third would be unreasonable.

**Review by darinkrauss — APPROVED** [review]
Nice, LGTM! 🎆

**Review by rickpasetto — APPROVED** [review]
LGTM 🧨
